wire-ios-data-model/Source/Core Crypto/CoreCryptoKeyProvider.swift

```diff
@@ -17,126 +17,272 @@
-        if let oldKey = try? fetchCoreCryptoKey() {
-            let item = CoreCryptoKeychainItem()
-            let newKey = try KeychainManager.generateKey(numberOfBytes: 32)
-            try await coreCryptoKeyMigrationManager?.updateKey(path: path, oldKey: oldKey, newKey: newKey)
+    private func logRotation(_ message: String) {
+        WireLogger.coreCrypto.info("[key rotation] \(message)", attributes: .safePublic)
+    }
 
+    private func deleteOrMarkAsStale(item: CoreCryptoKeychainItem) {
+        do {
             try KeychainManager.deleteItem(item)
-            try KeychainManager.storeItem(item, value: newKey)
+        } catch {
+            staleKeysTracker.addKey(id: item.uniqueKeyId)
         }
     }
 
-    private func migrateKeyIfNeeded(path: String) async throws {
-        WireLogger.coreCrypto.info("Migrating core crypto key...")
+    private func migrateDatabaseKeyToBytes(path: String) async throws {
+        guard coreCryptoKeyMigrationManager.isMigrationToBytesNeeded else { return }
 
-        if let oldKey = try? fetchCoreCryptoKey() {
+        // Getting the unscoped key, because if the scoped key exists, it's already in the right format.
+        if let oldKey = try? fetchCoreCryptoKey(scoped: false) {
             // Since version 6.x, CC has changed the key format and clients need to migrate the key.
             // We can reuse the same key, but the "new key" must be 'Data'.
-            try await coreCryptoKeyMigrationManager?.performMigrationIfNeeded(
-                path: path,
-                oldKey: oldKey.base64EncodedString(),
-                newKey: oldKey
-            )
+            do {
+                try await coreCryptoKeyMigrationManager.migrateDatabaseKeyToBytes(
+                    path: path,
+                    oldKey: oldKey.base64EncodedString(),
+                    newKey: oldKey
+                )
+            } catch {
+                throw Failure.failedToMigrateKeyToBytes(error)
+            }
         } else {
             // If there is no key,
             // then this is a fresh install and we do not need to perform migration.
-            coreCryptoKeyMigrationManager?.markMigrationAsSkipped()
+            coreCryptoKeyMigrationManager.markMigrationToBytesAsSkipped()
         }
     }
 
-    private func fetchCoreCryptoKey() throws -> Data {
-        let item = CoreCryptoKeychainItem()
-        return try KeychainManager.fetchItem(item)
+    private func fetchCoreCryptoKey(scoped: Bool = true) throws -> Data? {
+        do {
+            let item = CoreCryptoKeychainItem(
+                uniqueKeyId: uniqueKeyId,
+                userID: userID,
+                scoped: scoped
+            )
+            return try KeychainManager.fetchItem(item)
+        } catch KeychainManager.Error.failedToFetchItemFromKeychain(errSecItemNotFound) {
+            return nil
+        } catch {
+            throw error
+        }
```

Comment:
No need to catch and throw, if you omit this it'll be rethrown automatically.

```suggestion
        }
```

Reply:
[1f0a47f](https://github.com/wireapp/wire-ios/pull/3637/commits/1f0a47fdd9917d071734691e046f356d839e7cfd)